src/libs/route_labeler.js

```diff
@@ -0,0 +1,131 @@
+import { coordAll } from '@turf/meta';
+import { lineString } from '@turf/helpers';
+import { getCoord, getCoords } from '@turf/invariant';
+import along from '@turf/along';
+import lineLength from '@turf/length';
+import lineSliceAlong from '@turf/line-slice-along';
+import bearing from '@turf/bearing';
+import distance from '@turf/distance';
+
+// find the point at the given distance ratio on the linestring
+function project(ratio, ls) {
+  const length = lineLength(ls);
+  const lngLat = getCoord(along(ls, length * ratio));
+  // compute the approximate "axis" of the line (horizontal or vertical)
+  // around the label position to chose an anchor minimizing the portion of line covered.
+  const axis = getAxis(lineSliceAlong(ls, length * (ratio - 0.1), length * (ratio + 0.1)));
+
+  return { lngLat, axis };
+}
+
+function getAxis(ls) {
+  const segmentCoords = getCoords(ls);
+  const segmentBearing = bearing(segmentCoords[0], segmentCoords[segmentCoords.length - 1]);
+  if (Math.abs(segmentBearing) < 45 || Math.abs(segmentBearing) > 135) {
+    return 'vertical';
+  }
+  return 'horizontal';
+}
+
+const asKey = coord => `${coord[0].toFixed(6)},${coord[1].toFixed(6)}`;
+
+function distinctSegment(coordinates, coordCounts) {
+  let start = 0;
+  let end = coordinates.length - 1;
+
+  // a distinct segment is a part of a line where coordinates
+  // appear only once accross all the features
+  for (let index = 0; index < coordinates.length; index++) {
+    const coord = coordinates[index];
+    if (start === 0 && coordCounts.get(asKey(coord)) === 1) {
+      start = index;
+    }
+    if (start !== 0 && coordCounts.get(asKey(coord)) !== 1) {
+      end = index;
+      break;
+    }
+  }
+
+  return lineString(coordinates.slice(start, end));
+}
+
+// extract the first segment of each linestring
+// whose coordinates don't overlap with another feature
+export function findDistinctSegments(linestrings) {
+  if (linestrings.length < 2) {
+    return linestrings;
+  }
+  // extract raw coordinates
+  const featuresCoords = linestrings.map(ls => dropRepeatedCoords(coordAll(ls)));
+  // count occurences of each coordinate accross all features
+  const coordCounts = new Map();
+  [].concat(...featuresCoords).forEach(coord => {
```

Comment:
[...featuresCoords] :innocent: ?

Reply:
Nope, not the same, what I want is a `flat` operation, not a copy :)

Reply:
ah yes sorry, I misread. I guess we can't use`Array.prototype.flat()` because of IE ?

Reply:
That's it.